File: ls/server/src/test/java/org/opencds/cqf/cql/ls/server/repository/ig/standard/CqlContentTest.java
package org.opencds.cqf.cql.ls.server.repository.ig.standard;

import static org.junit.jupiter.api.Assertions.*;

import ca.uhn.fhir.rest.server.exceptions.ResourceNotFoundException;
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Path;
import org.hl7.fhir.dstu2.model.ValueSet;
import org.hl7.fhir.dstu3.model.Library;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import org.opencds.cqf.fhir.test.Resources;

class CqlContentTest {

    @TempDir
    static Path tempDir;

    @BeforeAll
    static void setup() throws ClassNotFoundException, URISyntaxException, IOException {
        // This copies the sample IG to a temporary directory so that
        // we can test against an actual filesystem
        Resources.copyFromJar("/sampleIgs/ig/standard/cqlContent", tempDir);
    }

    @Test
    void loadCqlContentDstu3() {
        var lib = new Library();
        lib.addContent().setContentType("text/cql").setUrl("cql/Test.cql");
        IgStandardCqlContent.loadCqlContent(lib, tempDir);
        assertNotNull(lib.getContentFirstRep().getData());
    }

    @Test
    void loadCqlContentR4() {
        var lib = new org.hl7.fhir.r4.model.Library();
        lib.addContent().setContentType("text/cql").setUrl("cql/Test.cql");
        IgStandardCqlContent.loadCqlContent(lib, tempDir);
        assertNotNull(lib.getContentFirstRep().getData());
    }

    @Test
    void loadCqlContentR5() {
        var lib = new org.hl7.fhir.r5.model.Library();
        lib.addContent().setContentType("text/cql").setUrl("cql/Test.cql");
        IgStandardCqlContent.loadCqlContent(lib, tempDir);
        assertNotNull(lib.getContentFirstRep().getData());
    }

    @Test
    void emptyLibraryDoesNothing() {
        var lib = new Library();
        IgStandardCqlContent.loadCqlContent(lib, tempDir);
        assertEquals(0, lib.getContent().size());
    }

    @Test
    void nonLibraryResourceDoesNotThrow() {
        assertDoesNotThrow(() -> IgStandardCqlContent.loadCqlContent(new ValueSet(), tempDir));
    }

    @Test
    void invalidFhirVersionThrows() {
        var lib = new org.hl7.fhir.r4b.model.Library();
        assertThrows(IllegalArgumentException.class, () -> IgStandardCqlContent.loadCqlContent(lib, tempDir));
    }

    @Test
    void invalidPathThrows() {
        var lib = new org.hl7.fhir.r4.model.Library();
        lib.addContent().setContentType("text/cql").setUrl("not-a-real-path/Test.cql");
        assertThrows(ResourceNotFoundException.class, () -> IgStandardCqlContent.loadCqlContent(lib, tempDir));
    }

    @Test
    void nullThrows() {
        assertThrows(NullPointerException.class, () -> IgStandardCqlContent.loadCqlContent(null, tempDir));

        var lib = new Library();
        assertThrows(NullPointerException.class, () -> IgStandardCqlContent.loadCqlContent(lib, null));
    }
}

File: ls/server/src/test/kotlin/org/opencds/cqf/cql/ls/server/repository/ig/standard/CqlContentTest.kt
package org.opencds.cqf.cql.ls.server.repository.ig.standard

import ca.uhn.fhir.rest.server.exceptions.ResourceNotFoundException
import java.nio.file.Path
import org.hl7.fhir.dstu2.model.ValueSet
import org.hl7.fhir.dstu3.model.Library
import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import org.opencds.cqf.fhir.test.Resources

class CqlContentTest {

    companion object {
        @TempDir
        @JvmField
        var tempDir: Path? = null

        @BeforeAll
        @JvmStatic
        fun setup() {
            // This copies the sample IG to a temporary directory so that
            // we can test against an actual filesystem
            Resources.copyFromJar("/sampleIgs/ig/standard/cqlContent", tempDir!!)
        }
    }

    @Test
    fun loadCqlContentDstu3() {
        val lib = Library()
        lib.addContent().setContentType("text/cql").url = "cql/Test.cql"
        IgStandardCqlContent.loadCqlContent(lib, tempDir!!)
        assertNotNull(lib.contentFirstRep.data)
    }

    @Test
    fun loadCqlContentR4() {
        val lib = org.hl7.fhir.r4.model.Library()
        lib.addContent().setContentType("text/cql").url = "cql/Test.cql"
        IgStandardCqlContent.loadCqlContent(lib, tempDir!!)
        assertNotNull(lib.contentFirstRep.data)
    }

    @Test
    fun loadCqlContentR5() {
        val lib = org.hl7.fhir.r5.model.Library()
        lib.addContent().setContentType("text/cql").url = "cql/Test.cql"
        IgStandardCqlContent.loadCqlContent(lib, tempDir!!)
        assertNotNull(lib.contentFirstRep.data)
    }

    @Test
    fun emptyLibraryDoesNothing() {
        val lib = Library()
        IgStandardCqlContent.loadCqlContent(lib, tempDir!!)
        assertEquals(0, lib.content.size)
    }

    @Test
    fun nonLibraryResourceDoesNotThrow() {
        assertDoesNotThrow { IgStandardCqlContent.loadCqlContent(ValueSet(), tempDir!!) }
    }

    @Test
    fun invalidFhirVersionThrows() {
        val lib = org.hl7.fhir.r4b.model.Library()
        assertThrows(IllegalArgumentException::class.java) { IgStandardCqlContent.loadCqlContent(lib, tempDir!!) }
    }

    @Test
    fun invalidPathThrows() {
        val lib = org.hl7.fhir.r4.model.Library()
        lib.addContent().setContentType("text/cql").url = "not-a-real-path/Test.cql"
        assertThrows(ResourceNotFoundException::class.java) { IgStandardCqlContent.loadCqlContent(lib, tempDir!!) }
    }

    @Suppress("ARGUMENT_TYPE_MISMATCH")
    @Test
    fun nullThrows() {
        @Suppress("ARGUMENT_TYPE_MISMATCH")
        assertThrows(NullPointerException::class.java) {
            @Suppress("ARGUMENT_TYPE_MISMATCH")
            IgStandardCqlContent.loadCqlContent(null as org.hl7.fhir.instance.model.api.IBaseResource?, tempDir!!)
        }

        val lib = Library()
        @Suppress("ARGUMENT_TYPE_MISMATCH")
        assertThrows(NullPointerException::class.java) {
            @Suppress("ARGUMENT_TYPE_MISMATCH")
            IgStandardCqlContent.loadCqlContent(lib, null as Path?)
        }
    }
}
